Mech_lab6/src/main/java/agh/ics/oop/AbstractWorldMap.java

```diff
@@ -0,0 +1,46 @@
+package agh.ics.oop;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+abstract public class AbstractWorldMap implements IWorldMap, IPositionChangeObserver {
+    private MapVisualizer map = new MapVisualizer(this);
+    protected Map<Vector2d, Animal> animals = new HashMap<>();
+    public abstract Vector2d getUpperRight();
+    public abstract Vector2d getLowerLeft();
+    abstract public boolean canMoveTo(Vector2d position);
+
+    public String toString() {
+        return this.map.draw(getLowerLeft(), getUpperRight());
+    }
+
+    public boolean isOccupied(Vector2d position) {
+        return (this.objectAt(position) != null);
+    }
+
+    public Object objectAt(Vector2d position) {
+        if (animals.containsKey(position)) {
+            return animals.get(position);
```

Comment:
można tu od razu zwrócić `animals.get(position)`, jeśli nie znajdzie to zwróci null